### DISCUSSION TITLE
errors during compilation on fresh arduino install with the libraries

### DISCUSSION THREAD

**Discussion by AyceBenz** [pschatzmann/arduino-audio-tools#1799] [Q&A] [answered]
/**
 * @file basic-a2dp-fft-led.ino
 * @brief A2DP Sink with output of the FFT result to the LED matrix   
 * For details see the FFT Wiki: https://github.com/pschatzmann/arduino-audio-tools/wiki/FFT
 * @author Phil Schatzmann
 * @copyright GPLv3
 */

#include "AudioTools.h"
#include "AudioTools/AudioLibs/AudioRealFFT.h" // or any other supported inplementation
#include "AudioTools/AudioLibs/LEDOutput.h"
#include "BluetoothA2DPSink.h"

#define PIN_LEDS 22
#define LED_X 8
#define LED_Y 1

BluetoothA2DPSink a2dp_sink;
AudioRealFFT fft; // or any other supported inplementation
FFTDisplay fft_dis(fft);
LEDOutput led(fft_dis); // output to LED matrix

// Provide data to FFT
void writeDataStream(const uint8_t *data, uint32_t length) {
  fft.write(data, length);
}

void setup() {
  Serial.begin(115200);
  AudioToolsLogger.begin(Serial, AudioToolsLogLevel::Info);

  // Setup FFT
  auto tcfg = fft.defaultConfig();
  tcfg.length = 1024;
  tcfg.channels = 2;
  tcfg.sample_rate = a2dp_sink.sample_rate();;
  tcfg.bits_per_sample = 16;
  fft.begin(tcfg);

  // Setup LED matrix output
  auto lcfg = led.defaultConfig();
  lcfg.x = LED_X;
  lcfg.y = LED_Y;
  led.begin(lcfg);

  fft_dis.fft_group_bin = 3;
  fft_dis.fft_start_bin = 0;
  fft_dis.fft_max_magnitude = 40000;
  fft_dis.begin();

  // add LEDs
  FastLED.addLeds<WS2812B, PIN_LEDS, GRB>(led.ledData(), led.ledCount());

  // register A2DP callback
  a2dp_sink.set_stream_reader(writeDataStream, false);

  // Start Bluetooth Audio Receiver
  Serial.print("starting a2dp-fft...");
  a2dp_sink.set_auto_reconnect(false);
  a2dp_sink.start("a2dp-fft");

}

void loop() { 
  led.update();
  delay(50);
}

**Comment by AyceBenz** [comment]
so this is the example code from the audiotools library but it wont work

**Comment by AyceBenz** [comment]
Arduino: 1.8.19 (Windows 10), Board: "ESP32 Dev Module, Disabled, Disabled, Huge APP (3MB No OTA/1MB SPIFFS), 240MHz (WiFi/BT), QIO, 80MHz, 4MB (32Mb), 921600, Core 1, Core 1, None, Disabled, Disabled"





















In file included from C:\Users\user\Documents\Arduino\libraries\arduino-audio-tools-main\src/AudioTools/AudioLibs/AudioRealFFT.h:4,

                 from C:\Users\user\AppData\Local\Temp\arduino_modified_sketch_982840\basic-a2dp-fft-led.ino:10:

C:\Users\user\Documents\Arduino\libraries\arduino-audio-tools-main\src/AudioTools/AudioLibs/FFT/FFTReal.h: In instantiation of 'class ffft::FFTRealFixLen<2>':

C:\Users\user\Documents\Arduino\libraries\arduino-audio-tools-main\src/AudioTools/AudioLibs/FFT/FFTReal.h:3735:30:   required from here

C:\Users\user\Documents\Arduino\libraries\arduino-audio-tools-main\src/AudioTools/AudioLibs/FFT/FFTReal.h:2440:82: warning: arithmetic between different enumeration types 'ffft::FFTRealFixLen<2>::<unnamed enum>' and 'ffft::FFTRealFixLen<2>::<unnamed enum>' is deprecated [-Wdeprecated-enum-enum-conversion]

 2440 |    enum {                       TRIGO_BD                        =   ((FFT_LEN_L2 - TRIGO_BD_LIMIT) < 0)

      |                                                                      ~~~~~~~~~~~~^~~~~~~~~~~~~~~~~

C:\Users\user\Documents\Arduino\libraries\arduino-audio-tools-main\src/AudioTools/AudioLibs/FFT/FFTReal.h:2446:78: warning: arithmetic between different enumeration types 'ffft::FFTRealFixLen<2>::<unnamed enum>' and 'ffft::FFTRealFixLen<2>::<unnamed enum>' is deprecated [-Wdeprecated-enum-enum-conversion]

 2446 |         enum {                  NBR_TRIGO_OSC                   = FFT_LEN_L2 - TRIGO_BD };

      |                                                                   ~~~~~~~~~~~^~~~~~~~~~

C:\Users\user\Documents\Arduino\libraries\arduino-audio-tools-main\src/AudioTools/AudioLibs/FFT/FFTReal.h: In instantiation of 'class ffft::FFTRealFixLen<1>':

C:\Users\user\Documents\Arduino\libraries\arduino-audio-tools-main\src/AudioTools/AudioLibs/FFT/FFTReal.h:3753:30:   required from here

C:\Users\user\Documents\Arduino\libraries\arduino-audio-tools-main\src/AudioTools/AudioLibs/FFT/FFTReal.h:2440:82: warning: arithmetic between different enumeration types 'ffft::FFTRealFixLen<1>::<unnamed enum>' and 'ffft::FFTRealFixLen<1>::<unnamed enum>' is deprecated [-Wdeprecated-enum-enum-conversion]

 2440 |    enum {                       TRIGO_BD                        =   ((FFT_LEN_L2 - TRIGO_BD_LIMIT) < 0)

      |                                                                      ~~~~~~~~~~~~^~~~~~~~~~~~~~~~~

C:\Users\user\Documents\Arduino\libraries\arduino-audio-tools-main\src/AudioTools/AudioLibs/FFT/FFTReal.h:2446:78: warning: arithmetic between different enumeration types 'ffft::FFTRealFixLen<1>::<unnamed enum>' and 'ffft::FFTRealFixLen<1>::<unnamed enum>' is deprecated [-Wdeprecated-enum-enum-conversion]

 2446 |         enum {                  NBR_TRIGO_OSC                   = FFT_LEN_L2 - TRIGO_BD };

      |                                                                   ~~~~~~~~~~~^~~~~~~~~~

C:\Users\user\Documents\Arduino\libraries\arduino-audio-tools-main\src/AudioTools/AudioLibs/FFT/FFTReal.h: In instantiation of 'class ffft::FFTRealFixLen<0>':

C:\Users\user\Documents\Arduino\libraries\arduino-audio-tools-main\src/AudioTools/AudioLibs/FFT/FFTReal.h:3765:30:   required from here

C:\Users\user\Documents\Arduino\libraries\arduino-audio-tools-main\src/AudioTools/AudioLibs/FFT/FFTReal.h:2440:82: warning: arithmetic between different enumeration types 'ffft::FFTRealFixLen<0>::<unnamed enum>' and 'ffft::FFTRealFixLen<0>::<unnamed enum>' is deprecated [-Wdeprecated-enum-enum-conversion]

 2440 |    enum {                       TRIGO_BD                        =   ((FFT_LEN_L2 - TRIGO_BD_LIMIT) < 0)

      |                                                                      ~~~~~~~~~~~~^~~~~~~~~~~~~~~~~

C:\Users\user\Documents\Arduino\libraries\arduino-audio-tools-main\src/AudioTools/AudioLibs/FFT/FFTReal.h:2446:78: warning: arithmetic between different enumeration types 'ffft::FFTRealFixLen<0>::<unnamed enum>' and 'ffft::FFTRealFixLen<0>::<unnamed enum>' is deprecated [-Wdeprecated-enum-enum-conversion]

 2446 |         enum {                  NBR_TRIGO_OSC                   = FFT_LEN_L2 - TRIGO_BD };

      |                                                                   ~~~~~~~~~~~^~~~~~~~~~

In file included from C:\Users\user\Documents\Arduino\libraries\FastLED\src/crgb.h:11,

                 from C:\Users\user\Documents\Arduino\libraries\FastLED\src/lib8tion/scale8.h:4,

                 from C:\Users\user\Documents\Arduino\libraries\FastLED\src/lib8tion/math8.h:3,

                 from C:\Users\user\Documents\Arduino\libraries\FastLED\src/lib8tion.h:377,

                 from C:\Users\user\Documents\Arduino\libraries\FastLED\src/pixeltypes.h:7,

                 from C:\Users\user\Documents\Arduino\libraries\FastLED\src/cpixel_ledcontroller.h:10,

                 from C:\Users\user\Documents\Arduino\libraries\FastLED\src/controller.h:7,

                 from C:\Users\user\Documents\Arduino\libraries\FastLED\src/FastLED.h:65,

                 from C:\Users\user\Documents\Arduino\libraries\arduino-audio-tools-main\src/AudioTools/AudioLibs/LEDOutput.h:2,

                 from C:\Users\user\AppData\Local\Temp\arduino_modified_sketch_982840\basic-a2dp-fft-led.ino:11:

C:\Users\user\Documents\Arduino\libraries\FastLED\src/template_magic.h:131:44: error: template argument 2 is invalid

  131 | typename fl::enable_if<fl::is_same<U, CLASS>::value, T&>::type \

      |                                            ^

C:\Users\user\Documents\Arduino\libraries\FastLED\src/str.h:248:1: note: in expansion of macro 'FASTLED_DEFINE_OUTPUT_OPERATOR'

  248 | FASTLED_DEFINE_OUTPUT_OPERATOR(Str) {

      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

C:\Users\user\Documents\Arduino\libraries\FastLED\src/template_magic.h:131:56: error: template argument 1 is invalid

  131 | typename fl::enable_if<fl::is_same<U, CLASS>::value, T&>::type \

      |                                                        ^

C:\Users\user\Documents\Arduino\libraries\FastLED\src/str.h:248:1: note: in expansion of macro 'FASTLED_DEFINE_OUTPUT_OPERATOR'

  248 | FASTLED_DEFINE_OUTPUT_OPERATOR(Str) {

      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

C:\Users\user\Documents\Arduino\libraries\FastLED\src/template_magic.h:131:57: error: '<declaration error>' is not a template [-fpermissive]

  131 | typename fl::enable_if<fl::is_same<U, CLASS>::value, T&>::type \

      |                                                         ^~

C:\Users\user\Documents\Arduino\libraries\FastLED\src/template_magic.h:131:57: note: in definition of macro 'FASTLED_DEFINE_OUTPUT_OPERATOR'

  131 | typename fl::enable_if<fl::is_same<U, CLASS>::value, T&>::type \

      |                                                         ^~

C:\Users\user\Documents\Arduino\libraries\FastLED\src/str.h:248:32: error: reference to 'Str' is ambiguous

  248 | FASTLED_DEFINE_OUTPUT_OPERATOR(Str) {

      |                                ^~~

C:\Users\user\Documents\Arduino\libraries\FastLED\src/template_magic.h:132:25: note: in definition of macro 'FASTLED_DEFINE_OUTPUT_OPERATOR'

  132 | operator<<(T& os, const CLASS& obj)

      |                         ^~~~~

In file included from C:\Users\user\Documents\Arduino\libraries\arduino-audio-tools-main\src/AudioTools/CoreAudio/AudioHttp/HttpHeader.h:4,

                 from C:\Users\user\Documents\Arduino\libraries\arduino-audio-tools-main\src/AudioTools/CoreAudio/AudioHttp/HttpChunkReader.h:3,

                 from C:\Users\user\Documents\Arduino\libraries\arduino-audio-tools-main\src/AudioTools/CoreAudio/AudioHttp/HttpRequest.h:5,

                 from C:\Users\user\Documents\Arduino\libraries\arduino-audio-tools-main\src/AudioTools/CoreAudio/AudioMetaData/MetaData.h:6,

                 from C:\Users\user\Documents\Arduino\libraries\arduino-audio-tools-main\src/AudioTools.h:97,

                 from C:\Users\user\AppData\Local\Temp\arduino_modified_sketch_982840\basic-a2dp-fft-led.ino:9:

C:\Users\user\Documents\Arduino\libraries\arduino-audio-tools-main\src/AudioTools/CoreAudio/AudioBasic/Str.h:23:7: note: candidates are: 'class audio_tools::Str'

   23 | class Str : public StrView {

      |       ^~~

In file included from C:\Users\user\Documents\Arduino\libraries\FastLED\src/screenmap.h:9,

                 from C:\Users\user\Documents\Arduino\libraries\FastLED\src/engine_events.h:6,

                 from C:\Users\user\Documents\Arduino\libraries\FastLED\src/cled_controller.h:17,

                 from C:\Users\user\Documents\Arduino\libraries\FastLED\src/cpixel_ledcontroller.h:16:

C:\Users\user\Documents\Arduino\libraries\FastLED\src/str.h:234:7: note:                 'class Str'

  234 | class Str : public StrN<FASTLED_STR_INLINED_SIZE> {

      |       ^~~

C:\Users\user\Documents\Arduino\libraries\FastLED\src/str.h: In function 'int operator<<(T&, const int&)':

C:\Users\user\Documents\Arduino\libraries\FastLED\src/str.h:249:15: error: request for member 'c_str' in 'obj', which is of non-class type 'const int'

  249 |     os << obj.c_str();

      |               ^~~~~

C:\Users\user\Documents\Arduino\libraries\FastLED\src/screenmap.h: At global scope:

C:\Users\user\Documents\Arduino\libraries\FastLED\src/screenmap.h:123:54: error: template argument 1 is invalid

  123 |                           FixedMap<Str, ScreenMap, 16> *segmentMaps);

      |                                                      ^

C:\Users\user\Documents\Arduino\libraries\FastLED\src/screenmap.h:123:54: error: template argument 3 is invalid

C:\Users\user\Documents\Arduino\libraries\FastLED\src/screenmap.h:123:54: error: template argument 1 is invalid

C:\Users\user\Documents\Arduino\libraries\FastLED\src/screenmap.h:123:54: error: template argument 3 is invalid

C:\Users\user\Documents\Arduino\libraries\FastLED\src/screenmap.h:123:54: error: template argument 1 is invalid

C:\Users\user\Documents\Arduino\libraries\FastLED\src/screenmap.h:123:54: error: template argument 3 is invalid

C:\Users\user\Documents\Arduino\libraries\FastLED\src/screenmap.h:123:54: error: template argument 1 is invalid

C:\Users\user\Documents\Arduino\libraries\FastLED\src/screenmap.h:123:54: error: template argument 3 is invalid

C:\Users\user\Documents\Arduino\libraries\FastLED\src/screenmap.h:123:27: error: invalid template-id

  123 |                           FixedMap<Str, ScreenMap, 16> *segmentMaps);

      |                           ^~~~~~~~

C:\Users\user\Documents\Arduino\libraries\FastLED\src/screenmap.h:123:36: error: reference to 'Str' is ambiguous

  123 |                           FixedMap<Str, ScreenMap, 16> *segmentMaps);

      |                                    ^~~

C:\Users\user\Documents\Arduino\libraries\arduino-audio-tools-main\src/AudioTools/CoreAudio/AudioBasic/Str.h:23:7: note: candidates are: 'class audio_tools::Str'

   23 | class Str : public StrView {

      |       ^~~

C:\Users\user\Documents\Arduino\libraries\FastLED\src/str.h:234:7: note:                 'class Str'

  234 | class Str : public StrN<FASTLED_STR_INLINED_SIZE> {

      |       ^~~

C:\Users\user\Documents\Arduino\libraries\FastLED\src/screenmap.h:123:27: error: template placeholder type 'FixedMap<...auto...>' must be followed by a simple declarator-id

  123 |                           FixedMap<Str, ScreenMap, 16> *segmentMaps);

      |                           ^~~~~~~~

In file included from C:\Users\user\Documents\Arduino\libraries\FastLED\src/screenmap.h:10:

C:\Users\user\Documents\Arduino\libraries\FastLED\src/fixed_map.h:17:7: note: 'template<class Key, class Value, unsigned int N> class FixedMap' declared here

   17 | class FixedMap {

      |       ^~~~~~~~

C:\Users\user\Documents\Arduino\libraries\FastLED\src/screenmap.h:125:60: error: template argument 1 is invalid

  125 |     static void toJsonStr(const FixedMap<Str, ScreenMap, 16>&, Str* jsonBuffer);

      |                                                            ^

C:\Users\user\Documents\Arduino\libraries\FastLED\src/screenmap.h:125:60: error: template argument 3 is invalid

C:\Users\user\Documents\Arduino\libraries\FastLED\src/screenmap.h:125:60: error: template argument 1 is invalid

C:\Users\user\Documents\Arduino\libraries\FastLED\src/screenmap.h:125:60: error: template argument 3 is invalid

C:\Users\user\Documents\Arduino\libraries\FastLED\src/screenmap.h:125:60: error: template argument 1 is invalid

C:\Users\user\Documents\Arduino\libraries\FastLED\src/screenmap.h:125:60: error: template argument 3 is invalid

C:\Users\user\Documents\Arduino\libraries\FastLED\src/screenmap.h:125:60: error: template argument 1 is invalid

C:\Users\user\Documents\Arduino\libraries\FastLED\src/screenmap.h:125:60: error: template argument 3 is invalid

C:\Users\user\Documents\Arduino\libraries\FastLED\src/screenmap.h:125:33: error: invalid template-id

  125 |     static void toJsonStr(const FixedMap<Str, ScreenMap, 16>&, Str* jsonBuffer);

      |                                 ^~~~~~~~

C:\Users\user\Documents\Arduino\libraries\FastLED\src/screenmap.h:125:42: error: reference to 'Str' is ambiguous

  125 |     static void toJsonStr(const FixedMap<Str, ScreenMap, 16>&, Str* jsonBuffer);

      |                                          ^~~

C:\Users\user\Documents\Arduino\libraries\arduino-audio-tools-main\src/AudioTools/CoreAudio/AudioBasic/Str.h:23:7: note: candidates are: 'class audio_tools::Str'

   23 | class Str : public StrView {

      |       ^~~

C:\Users\user\Documents\Arduino\libraries\FastLED\src/str.h:234:7: note:                 'class Str'

  234 | class Str : public StrN<FASTLED_STR_INLINED_SIZE> {

      |       ^~~

C:\Users\user\Documents\Arduino\libraries\FastLED\src/screenmap.h:125:27: error: class template placeholder 'FixedMap' not permitted in this context

  125 |     static void toJsonStr(const FixedMap<Str, ScreenMap, 16>&, Str* jsonBuffer);

      |                           ^~~~~

C:\Users\user\Documents\Arduino\libraries\FastLED\src/screenmap.h:125:27: note: use 'auto' for an abbreviated function template

C:\Users\user\Documents\Arduino\libraries\FastLED\src/screenmap.h:125:64: error: reference to 'Str' is ambiguous

  125 |     static void toJsonStr(const FixedMap<Str, ScreenMap, 16>&, Str* jsonBuffer);

      |                                                                ^~~

C:\Users\user\Documents\Arduino\libraries\arduino-audio-tools-main\src/AudioTools/CoreAudio/AudioBasic/Str.h:23:7: note: candidates are: 'class audio_tools::Str'

   23 | class Str : public StrView {

      |       ^~~

C:\Users\user\Documents\Arduino\libraries\FastLED\src/str.h:234:7: note:                 'class Str'

  234 | class Str : public StrN<FASTLED_STR_INLINED_SIZE> {

      |       ^~~

C:\Users\user\Documents\Arduino\libraries\FastLED\src/screenmap.h:125:64: error: 'Str' has not been declared

  125 |     static void toJsonStr(const FixedMap<Str, ScreenMap, 16>&, Str* jsonBuffer);

      |                                                                ^~~

C:\Users\user\Documents\Arduino\libraries\FastLED\src/screenmap.h:126:57: error: template argument 1 is invalid

  126 |     static void toJson(const FixedMap<Str, ScreenMap, 16>&, FLArduinoJson::JsonDocument* doc);

      |                                                         ^

C:\Users\user\Documents\Arduino\libraries\FastLED\src/screenmap.h:126:57: error: template argument 3 is invalid

C:\Users\user\Documents\Arduino\libraries\FastLED\src/screenmap.h:126:57: error: template argument 1 is invalid

C:\Users\user\Documents\Arduino\libraries\FastLED\src/screenmap.h:126:57: error: template argument 3 is invalid

C:\Users\user\Documents\Arduino\libraries\FastLED\src/screenmap.h:126:57: error: template argument 1 is invalid

C:\Users\user\Documents\Arduino\libraries\FastLED\src/screenmap.h:126:57: error: template argument 3 is invalid

C:\Users\user\Documents\Arduino\libraries\FastLED\src/screenmap.h:126:57: error: template argument 1 is invalid

C:\Users\user\Documents\Arduino\libraries\FastLED\src/screenmap.h:126:57: error: template argument 3 is invalid

C:\Users\user\Documents\Arduino\libraries\FastLED\src/screenmap.h:126:30: error: invalid template-id

  126 |     static void toJson(const FixedMap<Str, ScreenMap, 16>&, FLArduinoJson::JsonDocument* doc);

      |                              ^~~~~~~~

C:\Users\user\Documents\Arduino\libraries\FastLED\src/screenmap.h:126:39: error: reference to 'Str' is ambiguous

  126 |     static void toJson(const FixedMap<Str, ScreenMap, 16>&, FLArduinoJson::JsonDocument* doc);

      |                                       ^~~

C:\Users\user\Documents\Arduino\libraries\arduino-audio-tools-main\src/AudioTools/CoreAudio/AudioBasic/Str.h:23:7: note: candidates are: 'class audio_tools::Str'

   23 | class Str : public StrView {

      |       ^~~

C:\Users\user\Documents\Arduino\libraries\FastLED\src/str.h:234:7: note:                 'class Str'

  234 | class Str : public StrN<FASTLED_STR_INLINED_SIZE> {

      |       ^~~

C:\Users\user\Documents\Arduino\libraries\FastLED\src/screenmap.h:126:24: error: class template placeholder 'FixedMap' not permitted in this context

  126 |     static void toJson(const FixedMap<Str, ScreenMap, 16>&, FLArduinoJson::JsonDocument* doc);

      |                        ^~~~~

C:\Users\user\Documents\Arduino\libraries\FastLED\src/screenmap.h:126:24: note: use 'auto' for an abbreviated function template

Multiple libraries were found for "WiFi.h"

 Used: C:\Users\user\AppData\Local\Arduino15\packages\esp32\hardware\esp32\3.0.5\libraries\WiFi

 Not used: C:\Program Files (x86)\Arduino\libraries\WiFi

exit status 1

Error compiling for board ESP32 Dev Module.



This report would have more information with
"Show verbose output during compilation"
option enabled in File -> Preferences.

**Comment by pschatzmann** [comment, marked as answer]
It seems that FastLED introduced some breaking conflicts with 3.9
I suggest to use an old working version or push for proper [FastLED for namespace support](https://github.com/FastLED/FastLED/issues/1775) 

You can also prevent this error by making sure that you include this first:
#include <FastLED.h>
